src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/Main.java

```diff
@@ -407,37 +410,70 @@ private String compile(Path file, List<String> javacOpts, Context context) throw
         return mainClassName;
     }
 
+    private static boolean isPublic(Method method) {
+        return method != null && Modifier.isPublic(method.getModifiers());
+    }
+
+    private static boolean isStatic(Method method) {
+        return method != null && Modifier.isStatic(method.getModifiers());
+    }
+
     /**
      * Invokes the {@code main} method of a specified class, using a class loader that
      * will load recently compiled classes from memory.
      *
      * @param mainClassName the class to be executed
-     * @param appArgs the arguments for the {@code main} method
+     * @param mainArgs the arguments for the {@code main} method
      * @param context the context for the class to be executed
      * @throws Fault if there is a problem finding or invoking the {@code main} method
      * @throws InvocationTargetException if the {@code main} method throws an exception
      */
-    private void execute(String mainClassName, String[] appArgs, Context context)
+    private void execute(String mainClassName, String[] mainArgs, Context context)
             throws Fault, InvocationTargetException {
         System.setProperty("jdk.launcher.sourcefile", context.file.toString());
         ClassLoader cl = context.getClassLoader(ClassLoader.getSystemClassLoader());
         try {
             Class<?> appClass = Class.forName(mainClassName, true, cl);
-            Method main = appClass.getDeclaredMethod("main", String[].class);
-            int PUBLIC_STATIC = Modifier.PUBLIC | Modifier.STATIC;
-            if ((main.getModifiers() & PUBLIC_STATIC) != PUBLIC_STATIC) {
+            Method mainMethod = MainMethodFinder.findMainMethod(appClass);
+            int mods = mainMethod.getModifiers();
+            boolean isStatic = Modifier.isStatic(mods);
+            boolean isPublic = Modifier.isStatic(mods);
+            boolean noArgs = mainMethod.getParameterCount() == 0;
+
+            if (!PreviewFeatures.isEnabled() && (!isStatic || !isPublic)) {
                 throw new Fault(Errors.MainNotPublicStatic);
             }
-            if (!main.getReturnType().equals(void.class)) {
+
+            if (!mainMethod.getReturnType().equals(void.class)) {
                 throw new Fault(Errors.MainNotVoid);
             }
-            main.setAccessible(true);
-            main.invoke(0, (Object) appArgs);
+
+            // Similar to sun.launcher.LauncherHelper#executeMainClass
+            // but duplicated here to prevent additional launcher frames
+            mainMethod.setAccessible(true);
+
+            if (isStatic) {
+                if (noArgs) {
+                    mainMethod.invoke(appClass);
```

Comment:
Given that `Method.invoke(...)` on a static method ignores the `obj` param that's passed to it, perhaps pass `null`?

Reply:
Nostalgia and a useful reminder.